pulp_container/app/registry_api.py

```diff
@@ -413,48 +412,44 @@ def partial_update(self, request, path, pk=None):
 
         if whole:
             start = 0
-            end = chunk.size - 1
         else:
             content_range = request.META.get("HTTP_CONTENT_RANGE", "")
             match = self.content_range_pattern.match(content_range)
-            if not match:
-                start = 0
-                end = 0
-                chunk_size = 0
-            else:
-                start = int(match.group("start"))
-                end = int(match.group("end"))
-                chunk_size = end - start + 1
+            start = 0 if not match else int(match.group("start"))
 
         upload = get_object_or_404(models.Upload, repository=repository, pk=pk)
 
-        if upload.offset != start:
-            raise Exception
-        upload.append_chunk(chunk, chunk_size=chunk_size)
-        upload.save()
-        return UploadResponse(
-            upload=upload, path=path, content_length=upload.file.size, request=request
-        )
+        chunk = ContentFile(chunk.read())
+        with transaction.atomic():
+            if upload.size != start:
+                raise Exception
+
+            upload.append(chunk, upload.size)
+            upload.size += chunk.size
+            upload.save()
+
+        return UploadResponse(upload=upload, path=path, content_length=chunk.size, request=request)
 
     def put(self, request, path, pk=None):
-        """Handles creation of Uploads."""
+        """
+        Create a blob from uploaded chunks.
+        """
         _, repository = self.get_dr_push(request, path)
 
         digest = request.query_params["digest"]
         upload = models.Upload.objects.get(pk=pk, repository=repository)
+        chunks = UploadChunk.objects.filter(upload=upload).order_by("offset")
+
+        with NamedTemporaryFile("ab") as temp_file:
```

Comment:
Is there a `assemble_and_commit` method on core's Upload we could use here, now that we inherit it?

Reply:
Not at the moment. In pulpcore, we have the same code just for assembling uploaded chunks: https://github.com/pulp/pulpcore/blob/08d446ba1a72f1eaec8f437e15ff799ee6efbdc2/pulpcore/app/tasks/upload.py#L28-L34. I can implement only the method `assemble()` because the committing is not same in both scenarios. What do you think?

Reply:
Maybe later. Let's get this done first.